# MINI PROJECT 1 - TEAM 2
#### Anish Prakash
#### Naga Shiva Sai Ram Koushik Kanneti
#### Shehzadi Mahum Agha
#### Ju-Chin Lee

## PART 1: Installing Packages

In [4]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/yelp-dataset/Dataset_User_Agreement.pdf
/kaggle/input/yelp-dataset/yelp_academic_dataset_review.json
/kaggle/input/yelp-dataset/yelp_academic_dataset_checkin.json
/kaggle/input/yelp-dataset/yelp_academic_dataset_business.json
/kaggle/input/yelp-dataset/yelp_academic_dataset_tip.json
/kaggle/input/yelp-dataset/yelp_academic_dataset_user.json


In [5]:
from sqlalchemy import create_engine
!pip install PyMySql
import pymysql
import sqlalchemy

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 45.0/45.0 kB 1.6 MB/s eta 0:00:00


In [6]:
!pip install pyodbc
 
!apt-get update
!apt-get install -y unixodbc-dev
 
#install Microsoft ODBC Driver for SQL Server
!curl https://packages.microsoft.com/keys/microsoft.asc | apt-key add -
!curl https://packages.microsoft.com/config/ubuntu/20.04/prod.list | tee /etc/apt/sources.list.d/msprod.list
!apt-get update
!ACCEPT_EULA=Y apt-get install -y msodbcsql17

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 336.0/336.0 kB 6.5 MB/s eta 0:00:00:00:01
Get:1 https://packages.cloud.google.com/apt gcsfuse-focal InRelease [1227 B]
Get:2 http://security.ubuntu.com/ubuntu focal-security InRelease [128 kB]      
Hit:3 http://archive.ubuntu.com/ubuntu focal InRelease                         
Get:4 http://archive.ubuntu.com/ubuntu focal-updates InRelease [128 kB]        
Get:5 https://packages.cloud.google.com/apt cloud-sdk InRelease [1618 B]       
Get:6 https://packages.cloud.google.com/apt gcsfuse-focal/main amd64 Packages [28.6 kB]
Get:7 http://archive.ubuntu.com/ubuntu focal-backports InRelease [128 kB]
Get:8 https://packages.cloud.google.com/apt cloud-sdk/main all Packages [1564 kB]
Get:9 http://security.ubuntu.com/ubuntu focal-security/restricted amd64 Packages [4105 kB]
Get:10 https://packages.cloud.google.com/apt cloud-sdk/main amd64 Packages [3375 kB]
Get:11 http://archive.ubuntu.com/ubuntu focal-updates/multiverse amd64 Packages [33.5 kB]
Get:12 

## PART 2: Securing Connection

In [8]:
!curl ifconfig.me #to know the IP

34.83.39.167

In [9]:
from sqlalchemy import create_engine
import pyodbc
 
# Create SQLAlchemy engine
username = 'anishp5admin'
password = 'anish&2003'
server = 'anishp5-server.database.windows.net'
database = 'MiniProject-1'
driver= 'ODBC Driver 17 for SQL Server'
connection_string = f"mssql+pyodbc://{username}:{password}@{server}/{database}?driver=ODBC+Driver+17+for+SQL+Server&connect_timeout=30"
engine = create_engine(connection_string)


#Test connection to Azure
from sqlalchemy import text
try:
    with engine.connect() as connection:
        result = connection.execute(text("SELECT 1;"))
        print("Connection Success, Return result :", result.fetchone())
except Exception as e:
    print("Connection Failed:", str(e))

Connection Success, Return result : (1,)


## PART 3: Data Exploration

In [10]:
review_df = pd.read_json('/kaggle/input/yelp-dataset/yelp_academic_dataset_review.json', lines = True, nrows = 25000)
business_df = pd.read_json('/kaggle/input/yelp-dataset/yelp_academic_dataset_business.json', lines = True, nrows = 25000)
checkin_df = pd.read_json('/kaggle/input/yelp-dataset/yelp_academic_dataset_checkin.json', lines = True, nrows = 25000)
user_df = pd.read_json('/kaggle/input/yelp-dataset/yelp_academic_dataset_user.json', lines = True, nrows = 25000)
tip_df = pd.read_json('/kaggle/input/yelp-dataset/yelp_academic_dataset_tip.json', lines = True, nrows = 25000)

In [11]:
business_df.info()
business_df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25000 entries, 0 to 24999
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   business_id   25000 non-null  object 
 1   name          25000 non-null  object 
 2   address       25000 non-null  object 
 3   city          25000 non-null  object 
 4   state         25000 non-null  object 
 5   postal_code   25000 non-null  object 
 6   latitude      25000 non-null  float64
 7   longitude     25000 non-null  float64
 8   stars         25000 non-null  float64
 9   review_count  25000 non-null  int64  
 10  is_open       25000 non-null  int64  
 11  attributes    22784 non-null  object 
 12  categories    24982 non-null  object 
 13  hours         21156 non-null  object 
dtypes: float64(3), int64(2), object(9)
memory usage: 2.7+ MB


business_id                      name  \
0  Pns2l4eNsfO8kk83dixA6A  Abby Rappoport, LAC, CMQ   
1  mpf3x-BjTdTEA3yCZrAYPw             The UPS Store   
2  tUFrWirKiKi_TAnsVWINQQ                    Target   
3  MTSW4McQd7CbVtyjqoe9mw        St Honore Pastries   
4  mWMc6_wTdE0EUBKIGXDVfA  Perkiomen Valley Brewery   

                           address           city state postal_code  \
0           1616 Chapala St, Ste 2  Santa Barbara    CA       93101   
1  87 Grasso Plaza Shopping Center         Affton    MO       63123   
2             5255 E Broadway Blvd         Tucson    AZ       85711   
3                      935 Race St   Philadelphia    PA       19107   
4                    101 Walnut St     Green Lane    PA       18054   

    latitude   longitude  stars  review_count  is_open  \
0  34.426679 -119.711197    5.0             7        0   
1  38.551126  -90.335695    3.0            15        1   
2  32.223236 -110.880452    3.5            22        0   
3  39.955505  -75.155564    4.0            80        1   
4  40.338183  -75.471659    4.5            13        1   

                                          attributes  \
0                      {'ByAppointmentOnly': 'True'}   
1             {'BusinessAcceptsCreditCards': 'True'}   
2  {'BikeParking': 'True', 'BusinessAcceptsCredit...   
3  {'RestaurantsDelivery': 'False', 'OutdoorSeati...   
4  {'BusinessAcceptsCreditCards': 'True', 'Wheelc...   

                                          categories  \
0  Doctors, Traditional Chinese Medicine, Naturop...   
1  Shipping Centers, Local Services, Notaries, Ma...   
2  Department Stores, Shopping, Fashion, Home & G...   
3  Restaurants, Food, Bubble Tea, Coffee & Tea, B...   
4                          Brewpubs, Breweries, Food   

                                               hours  
0                                               None  
1  {'Monday': '0:0-0:0', 'Tuesday': '8:0-18:30', ...  
2  {'Monday': '8:0-22:0', 'Tuesday': '8:0-22:0', ...  
3  {'Monday': '7:0-20:0', 'Tuesday': '7:0-20:0', ...  
4  {'Wednesday': '14:0-22:0', 'Thursday': '16:0-2...

### 3.1 Scope of Business Data:
This dataset encompasses businesses across various categories and locations within the Yelp platform, providing a comprehensive overview of the services offered. It includes information from diverse industries such as restaurants, retail stores, service providers, and more. The dataset is useful for analyzing business performance, customer feedback, and geographical distribution, enabling insights into trends and patterns within the Yelp ecosystem.

This dataset serves as a valuable resource for exploring business characteristics, customer engagement through reviews, and operational hours, contributing to a deeper understanding of consumer preferences and behaviors.

This dataset contains metadata for businesses listed on Yelp. It includes the following columns:

- **business_id**: A unique identifier for each business, represented as a string.
- **name**: The name of the business as a string.
- **address**: The physical address of the business as a string.
- **city**: The city where the business is located, represented as a string.
- **state**: The state where the business is located, represented as a string.
- **postal_code**: The postal code for the business's location, represented as a string.
- **latitude**: A float value indicating the geographic latitude of the business.
- **longitude**: A float value indicating the geographic longitude of the business.
- **stars**: A float representing the average star rating given to the business.
- **review_count**: An integer denoting the total number of reviews written for the business.
- **is_open**: An integer indicating whether the business is currently open (1) or closed (0).
- **attributes**: A JSON-like object that contains various attributes of the business, such as payment methods or special features.
- **categories**: A string listing the categories that the business falls under, separated by commas.
- **hours**: A JSON-like object detailing the operating hours of the business for each day of the week.

In [12]:
review_df.info()
review_df.head()
review_df.describe()
earliest_date = review_df['date'].min()
latest_date = review_df['date'].max()

print("Earliest date:", earliest_date)
print("Latest date:", latest_date)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25000 entries, 0 to 24999
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   review_id    25000 non-null  object        
 1   user_id      25000 non-null  object        
 2   business_id  25000 non-null  object        
 3   stars        25000 non-null  int64         
 4   useful       25000 non-null  int64         
 5   funny        25000 non-null  int64         
 6   cool         25000 non-null  int64         
 7   text         25000 non-null  object        
 8   date         25000 non-null  datetime64[ns]
dtypes: datetime64[ns](1), int64(4), object(4)
memory usage: 1.7+ MB
Earliest date: 2005-03-01 17:47:15
Latest date: 2018-10-04 18:22:35


### 3.2 Scope of Review Data:
This dataset consists of reviews from the Yelp platform, offering insights into customer feedback and engagement. It includes the following columns:

- **review_id**: A unique identifier for each review, represented as a string.
- **user_id**: The unique identifier of the user who wrote the review, represented as a string.
- **business_id**: The unique identifier of the business being reviewed, represented as a string.
- **stars**: The rating given by the user, ranging from 1 to 5, represented as an integer.
- **useful**: The number of times the review was marked as useful, represented as an integer.
- **funny**: The number of times the review was marked as funny, represented as an integer.
- **cool**: The number of times the review was marked as cool, represented as an integer.
- **text**: The written content of the review, represented as a string.
- **date**: The date and time the review was posted, represented as a datetime.




In [13]:
checkin_df.info()
checkin_df.head()
checkin_df.describe()

earliest_date = str(checkin_df['date'].min())[:10]
latest_date = checkin_df['date'].max()

print("Earliest date:", earliest_date)
print("Latest date:", latest_date)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25000 entries, 0 to 24999
Data columns (total 2 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   business_id  25000 non-null  object
 1   date         25000 non-null  object
dtypes: object(2)
memory usage: 390.8+ KB
Earliest date: 2010-01-16
Latest date: 2022-01-16 17:10:41, 2022-01-16 18:58:33, 2022-01-16 19:22:04, 2022-01-16 19:43:33, 2022-01-16 20:01:52, 2022-01-16 20:29:27, 2022-01-16 20:31:29, 2022-01-16 21:06:26, 2022-01-16 21:10:47, 2022-01-16 21:14:41, 2022-01-16 22:07:24, 2022-01-16 22:09:19, 2022-01-16 22:27:26, 2022-01-16 22:45:06, 2022-01-17 16:49:02, 2022-01-17 18:08:02, 2022-01-17 18:54:10, 2022-01-17 18:55:36, 2022-01-17 19:22:22, 2022-01-17 19:33:26, 2022-01-17 20:26:31, 2022-01-17 20:41:14, 2022-01-17 21:17:21, 2022-01-17 21:36:59, 2022-01-17 21:39:36, 2022-01-17 23:00:28, 2022-01-18 00:56:35, 2022-01-18 01:13:31, 2022-01-18 19:53:52, 2022-01-18 20:50:49, 2022-01-18 22:38:

### 3.3 Scope of Checkin Dataset:
The dataset consists of 2 columns and 131,930 entries, focusing on user engagement through check-ins at businesses. While it provides valuable insights into user activity and popular times for specific locations.

- **business_id**: A unique identifier for the business, linking to the business dataset.
- **date**: A string containing the timestamps of check-ins, formatted as a comma-separated list of datetime strings.

In [14]:
tip_df.info()
tip_df.head()

tip_df.describe()
earliest_date = tip_df['date'].min()
latest_date = tip_df['date'].max()

print("Earliest date:", earliest_date)
print("Latest date:", latest_date)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25000 entries, 0 to 24999
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   user_id           25000 non-null  object        
 1   business_id       25000 non-null  object        
 2   text              25000 non-null  object        
 3   date              25000 non-null  datetime64[ns]
 4   compliment_count  25000 non-null  int64         
dtypes: datetime64[ns](1), int64(1), object(3)
memory usage: 976.7+ KB
Earliest date: 2009-04-18 14:16:45
Latest date: 2018-05-05 08:19:29


### 3.4 Scope of Tip Dataset:
The dataset consists of 5 columns and 908,915 entries, focusing on user insights and recommendations for businesses. It offers valuable qualitative data about user experiences, but it does not include comprehensive user profiles or detailed information about the businesses themselves. Analyzing this dataset in conjunction with the business and user datasets could enhance understanding of customer sentiments and behaviors on the Yelp platform.

- **user_id**: A unique identifier for the user who made the tip.
- **business_id**: A unique identifier for the business associated with the tip, linking to the business dataset.
- **text**: The content of the tip, capturing user opinions or suggestions.
- **date**: The timestamp indicating when the tip was made, formatted as a datetime object.
- **compliment_count**: A count of how many compliments this tip has received.

In [15]:
user_df.info()
user_df.head()
user_df.describe()
earliest_date = user_df['yelping_since'].min()
latest_date = user_df['yelping_since'].max()

print("Earliest date:", earliest_date)
print("Latest date:", latest_date)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25000 entries, 0 to 24999
Data columns (total 22 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   user_id             25000 non-null  object 
 1   name                25000 non-null  object 
 2   review_count        25000 non-null  int64  
 3   yelping_since       25000 non-null  object 
 4   useful              25000 non-null  int64  
 5   funny               25000 non-null  int64  
 6   cool                25000 non-null  int64  
 7   elite               25000 non-null  object 
 8   friends             25000 non-null  object 
 9   fans                25000 non-null  int64  
 10  average_stars       25000 non-null  float64
 11  compliment_hot      25000 non-null  int64  
 12  compliment_more     25000 non-null  int64  
 13  compliment_profile  25000 non-null  int64  
 14  compliment_cute     25000 non-null  int64  
 15  compliment_list     25000 non-null  int64  
 16  comp

### 3.5 Scope of User Dataset:
This dataset encompasses 22 columns and 1,987,897 rows, focusing on user engagement and social interaction within the Yelp platform. Key metrics include `review_count`, `fans`, and various `compliments` received, providing a comprehensive view of user activity and contributions. Additionally, the `yelping_since` column captures the duration of user engagement on Yelp. The dataset offers insights into user behavior and popularity.

- **user_id**: A unique 22-character string identifying the user in `user.json`.
- **name**: The first name of the user as a string.
- **review_count**: An integer indicating the total number of reviews written by the user.
- **yelping_since**: A string representing the date the user joined Yelp, formatted as `YYYY-MM-DD`.
- **friends**: An array of strings listing the `user_ids` of the user's friends.
- **useful**: An integer reflecting the number of useful votes received on the user's reviews.
- **funny**: An integer indicating the number of funny votes the user's reviews have received.
- **cool**: An integer showing the number of cool votes garnered by the user's reviews.
- **fans**: An integer representing the total number of fans that the user has on Yelp.
- **elite**: An array of integers indicating the years during which the user was recognized as an elite member.
- **average_stars**: A float value representing the average star rating given by the user across all their reviews.
- **compliment_hot**: An integer quantifying the number of hot compliments received by the user.
- **compliment_more**: An integer representing the number of more compliments received by the user.
- **compliment_profile**: An integer denoting the number of profile-related compliments received by the user.
- **compliment_cute**: An integer for the number of cute compliments the user has received.
- **compliment_list**: An integer indicating the number of list-related compliments received by the user.
- **compliment_note**: An integer reflecting the number of note-related compliments received by the user.
- **compliment_plain**: An integer representing the number of plain compliments received by the user.
- **compliment_cool**: An integer denoting the number of cool compliments the user has received.
- **compliment_funny**: An integer indicating the number of funny compliments received by the user.
- **compliment_writer**: An integer showing the number of writer-related compliments the user has received.
- **compliment_photos**: An integer for the number of photo-related compliments received by the user.



## PART 4: PORTING DATA TO AZURE

In [16]:
#Test connection to Azure
from sqlalchemy import text
try:
    with engine.connect() as connection:
        result = connection.execute(text("SELECT 1;"))
        print("Connection Success, Return result :", result.fetchone())
except Exception as e:
    print("Connection Failed:", str(e))

Connection Success, Return result : (1,)


In [17]:
business_df.to_sql(name = "business", con = engine, if_exists='replace', index=True, dtype={'attributes': sqlalchemy.types.JSON, 'hours': sqlalchemy.types.JSON})

119

In [18]:
checkin_df.to_sql(name = "checkin", con = engine, if_exists='replace', index=True)

535

In [19]:
review_df.to_sql(name = "review", con = engine, if_exists='replace', index=True)

129

In [20]:
tip_df.to_sql(name = "tip", con = engine, if_exists='replace', index=True)

221

In [21]:
user_df.to_sql(name = "users", con = engine, if_exists='replace', index=True)

66

In [22]:
from sqlalchemy import create_engine

%load_ext sql

# Create SQLAlchemy engine
username = 'anishp5admin'
password = 'anish&2003'
server = 'anishp5-server.database.windows.net'
database = 'MiniProject-1'
driver= 'ODBC Driver 17 for SQL Server'
connection_string = f"mssql+pyodbc://{username}:{password}@{server}/{database}?driver=ODBC+Driver+17+for+SQL+Server&connect_timeout=30"
engine = create_engine(connection_string)


#Test connection to Azure
from sqlalchemy import text
try:
    with engine.connect() as connection:
        result = connection.execute(text("SELECT 1;"))
        print("Connection Success, Return result :", result.fetchone())
except Exception as e:
    print("Connection Failed:", str(e))


The sql extension is already loaded. To reload it, use:
  %reload_ext sql
Connection Success, Return result : (1,)


In [23]:
%load_ext sql

%sql $connection_string
 

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


### 4.1: Handling JSON Columns in Azure SQL Server

In [24]:
%%sql
 
-- Drop existing primary keys from all tables
 
ALTER TABLE business DROP CONSTRAINT IF EXISTS PK_business;
ALTER TABLE checkin DROP CONSTRAINT IF EXISTS PK_checkin;
ALTER TABLE review DROP CONSTRAINT IF EXISTS PK_review;
ALTER TABLE tip DROP CONSTRAINT IF EXISTS PK_tip;
ALTER TABLE users DROP CONSTRAINT IF EXISTS PK_user; 
 
-- Change the column length in order to setup primary keys
 
ALTER TABLE business ALTER COLUMN business_id VARCHAR(22) NOT NULL;
ALTER TABLE review ALTER COLUMN review_id VARCHAR(22) NOT NULL;
ALTER TABLE tip ALTER COLUMN tip_id INT NOT NULL;
ALTER TABLE users ALTER COLUMN user_id VARCHAR(22) NOT NULL;
 
-- Set the primary key for all tables
 
ALTER TABLE business ADD CONSTRAINT PK_business PRIMARY KEY (business_id);
ALTER TABLE review ADD CONSTRAINT PK_review PRIMARY KEY (review_id);
ALTER TABLE tip ADD CONSTRAINT PK_tip PRIMARY KEY (tip_id);
ALTER TABLE users ADD CONSTRAINT PK_user PRIMARY KEY (user_id);

 * mssql+pyodbc://anishp5admin:***@anishp5-server.database.windows.net/MiniProject-1?connect_timeout=30&driver=ODBC+Driver+17+for+SQL+Server
Done.
Done.
Done.
Done.
Done.
Done.
Done.
(pyodbc.ProgrammingError) ('42S22', "[42S22] [Microsoft][ODBC Driver 17 for SQL Server][SQL Server]ALTER TABLE ALTER COLUMN failed because column 'tip_id' does not exist in table 'tip'. (4924) (SQLExecDirectW)")
[SQL: ALTER TABLE tip ALTER COLUMN tip_id INT NOT NULL;]
(Background on this error at: https://sqlalche.me/e/20/f405)


### 4.2: Verification through Sample Query

In [25]:
%%sql
SELECT TOP 10 *
FROM review;

 * mssql+pyodbc://anishp5admin:***@anishp5-server.database.windows.net/MiniProject-1?connect_timeout=30&driver=ODBC+Driver+17+for+SQL+Server
Done.


[(0, 'KU_O5udG6zpxOg-VcAEodg', 'mh_-eMZ6K5RLWhZyISBhwA', 'XQfwVwDr-v0ZS3_CbbE5Xw', 3, 0, 0, 0, "If you decide to eat here, just be aware it is going to take about 2 hours from beginning to end. We have tried it multiple times, because I want to  ... (217 characters truncated) ... oo many experiences where we spent way too long waiting. We usually opt for another diner or restaurant on the weekends, in order to be done quicker.", datetime.datetime(2018, 7, 7, 22, 9, 11)),
 (1, 'BiTunyQ73aT9WBnpR9DZGw', 'OyoGAe7OKpv6SyGZT5g77Q', '7ATYjTIgM3jUlt4UM3IypQ', 5, 1, 0, 1, "I've taken a lot of spin classes over the years, and nothing compares to the classes at Body Cycle. From the nice, clean space and amazing bikes, to  ... (535 characters truncated) ... uragement, and is open to ideas and recommendations from anyone. Russell always wears a smile on his face, even when he's kicking your butt in class!", datetime.datetime(2012, 1, 3, 15, 28, 18)),
 (2, 'saUsX_uimxRlCVr67Z4Jig', '8g_iMtfSiwikVnbP2etR0A', 'YjUWPpI6HXG530lwP-fb2A', 3, 0, 0, 0, 'Family diner. Had the buffet. Eclectic assortment: a large chicken leg, fried jalapeño, tamale, two rolled grape leaves, fresh melon. All good. Lots  ... (41 characters truncated) ...  with breakfast served all day long. Friendly, attentive staff. Good place for a casual relaxed meal with no expectations. Next to the Clarion Hotel.', datetime.datetime(2014, 2, 5, 20, 30, 30)),
 (3, 'AqPFMleE6RsU23_auESxiA', '_7bHUi9Uuf5__HHc_Q8guQ', 'kxX2SOes4o-D3ZQBkiMRfA', 5, 1, 0, 1, "Wow!  Yummy, different,  delicious.   Our favorite is the lamb curry and korma.  With 10 different kinds of naan!!!  Don't let the outside deter you (because we almost changed our minds)...go in and try something new!   You'll be glad you did!", datetime.datetime(2015, 1, 4, 0, 1, 3)),
 (4, 'Sx8TMOWLNuJBWer-0pcmoA', 'bcjbaE6dDog4jkNY91ncLQ', 'e4Vwtrqf-wpJfwesgvdgxQ', 4, 1, 0, 1, "Cute interior and owner (?) gave us tour of upcoming patio/rooftop area which will be great on beautiful days like today. Cheese curds were very good ... (236 characters truncated) ... ore menu options added such as salads w fun cheeses. Lots of beer and wine as well as limited cocktails. Next time I will try one of the draft wines.", datetime.datetime(2017, 1, 14, 20, 54, 15)),
 (5, 'JrIxlS1TzJ-iCu79ul40cQ', 'eUta8W_HdHMXPzLBBZhL1A', '04UD14gamNjLY0IDYVhHJg', 1, 1, 2, 1, "I am a long term frequent customer of this establishment. I just went in to order take out (3 apps) and was told they're too busy to do it. Really? T ... (43 characters truncated) ... our dick reach your ass? Yes? Go fuck yourself! I'm a frequent customer AND great tipper. Glad that Kanella just opened. NEVER going back to dmitris!", datetime.datetime(2015, 9, 23, 23, 10, 31)),
 (6, '6AxgBCNX_PNTOxmbRSwcKQ', 'r3zeYsv1XFBRA4dJpL78cw', 'gmjsEdUsKpj9Xxu6pdjH0g', 5, 0, 2, 0, "Loved this tour! I grabbed a groupon and the price was great. It was the perfect way to explore New Orleans for someone who'd never been there before ... (506 characters truncated) ... er), and you'll stop at a visitor center part way through the tour for a bathroom break if needed. This tour was one of my favorite parts of my trip!", datetime.datetime(2015, 1, 3, 23, 21, 18)),
 (7, '_ZeMknuYdlQcUqng_Im3yg', 'yfFzsLmaWF2d4Sr0UNbBgg', 'LHSTtnW3YHCeUkRDGyJOyw', 5, 2, 0, 0, 'Amazingly amazing wings and homemade bleu cheese. Had the ribeye: tender, perfectly prepared, delicious. Nice selection of craft beers. Would DEFINITELY recommend checking out this hidden gem.', datetime.datetime(2015, 8, 7, 2, 29, 16)),
 (8, 'ZKvDG2sBvHVdF5oBNUOpAQ', 'wSTuiTk-sKNdcFyprzZAjg', 'B5XSoSG3SfvQGtKEGQ1tSQ', 3, 1, 1, 0, "This easter instead of going to Lopez Lake we went to Los Padres National Forest which is really pretty but if you go to white rock the staff needs t ... (228 characters truncated) ... rest staff need to work hard to maintain this forest looking pretty and not like a dumpster. Even Cachuma lake looks like they put a bit more

## PART 5: 10 SQL queries

### 5.1: Top 5 cities with most restaurants
#### This query helps us see which cities have the most restaurant-related businesses, giving insights into competitive density and potential markets.

In [26]:
%%sql
SELECT TOP 5 city, COUNT(*) AS num_restaurants
FROM business
WHERE categories LIKE '%Restaurant%'
GROUP BY city
ORDER BY num_restaurants DESC;



 * mssql+pyodbc://anishp5admin:***@anishp5-server.database.windows.net/MiniProject-1?connect_timeout=30&driver=ODBC+Driver+17+for+SQL+Server
Done.


[('Philadelphia', 994),
 ('Tampa', 498),
 ('Indianapolis', 471),
 ('Nashville', 433),
 ('Tucson', 398)]

### 5.2 Find Average Star Rating of Restaurants in the City
#### Provides insight into the general quality and customer satisfaction level of restaurants in each of the top 5 cities.

In [29]:
%%sql
SELECT city, ROUND(AVG(stars),2) AS avg_rating
FROM business
WHERE city IN ('Tampa', 'Philadelphia', 'Indianapolis', 'Nashville', 'Tucson') 
  AND categories LIKE '%Restaurant%'
GROUP BY city
ORDER BY avg_rating DESC;


 * mssql+pyodbc://anishp5admin:***@anishp5-server.database.windows.net/MiniProject-1?connect_timeout=30&driver=ODBC+Driver+17+for+SQL+Server
Done.


[('Tampa', 3.67),
 ('Nashville', 3.57),
 ('Philadelphia', 3.56),
 ('Indianapolis', 3.46),
 ('Tucson', 3.44)]

### 5.3: Average Number of Fans for Yelp Users in the City
#### Helps gauge the general engagement and influence of Yelp users in the top 5 city, which is valuable for planning marketing and making restaurant choice.

In [28]:
%%sql
SELECT business.city, AVG(users.fans) AS avg_fans
FROM users
JOIN review ON users.user_id = review.user_id
JOIN business ON review.business_id = business.business_id
WHERE business.city IN ('Tampa', 'Philadelphia', 'Indianapolis', 'Nashville', 'Tucson')
GROUP BY business.city
ORDER BY avg_fans DESC;

 * mssql+pyodbc://anishp5admin:***@anishp5-server.database.windows.net/MiniProject-1?connect_timeout=30&driver=ODBC+Driver+17+for+SQL+Server
Done.


[('Indianapolis', 51),
 ('Tampa', 39),
 ('Philadelphia', 36),
 ('Nashville', 33),
 ('Tucson', 22)]

### 5.4: Find Categories with the Most Reviews in a City
#### Reveals popular restaurant categories in a chosen city, giving insights into the most sought-after cuisines.

In [30]:
%%sql
SELECT TOP 10 categories, SUM(review_count) AS total_reviews
FROM BUSINESS
WHERE city = 'Tampa'
GROUP BY categories
ORDER BY total_reviews DESC;


 * mssql+pyodbc://anishp5admin:***@anishp5-server.database.windows.net/MiniProject-1?connect_timeout=30&driver=ODBC+Driver+17+for+SQL+Server
Done.


[('Seafood, Japanese, Nightlife, Dance Clubs, Bars, American (New), Restaurants, Sushi Bars', 1115),
 ('Mexican, Restaurants, Salad, Food, Bars, Nightlife', 1092),
 ('American (Traditional), Restaurants, Venues & Event Spaces, Event Planning & Services, Breakfast & Brunch, Bars, Nightlife', 768),
 ('Restaurants, Caribbean', 746),
 ('African, Venues & Event Spaces, Steakhouses, Restaurants, Seafood, American (New), Event Planning & Services', 745),
 ('Food, Hotels, Coffee & Tea, Hotels & Travel, Restaurants, Casinos, Burgers, Bars, Nightlife, Arts & Entertainment, Event Planning & Services', 739),
 ('Golf, Sports Bars, Nightlife, Active Life, Bars, Restaurants, American (New)', 734),
 ('Restaurants, Burgers, American (Traditional), Bars, Pubs, Nightlife', 561),
 ('Bars, Breakfast & Brunch, Restaurants, American (New), Coffee & Tea, Food, Nightlife, Vegetarian', 548),
 ('Wine Bars, Desserts, Bars, American (New), Nightlife, Restaurants, Italian, Food, Salad, Breakfast & Brunch, Pizza', 544)]

### 5.5: Identify Popular Restaurant Categories in the City
#### Helps identify popular restaurant categories in the city, allowing us to understand what types of restaurants are common and possibly find a niche.

In [31]:
%%sql
SELECT TOP 5 categories, COUNT(*) AS num_businesses
FROM business
WHERE city = 'Tampa' AND categories LIKE '%Restaurant%'
GROUP BY categories
ORDER BY num_businesses DESC;



 * mssql+pyodbc://anishp5admin:***@anishp5-server.database.windows.net/MiniProject-1?connect_timeout=30&driver=ODBC+Driver+17+for+SQL+Server
Done.


[('Restaurants, Chinese', 7),
 ('Pizza, Restaurants', 5),
 ('Chinese, Restaurants', 4),
 ('Restaurants, Pizza', 4),
 ('Restaurants, Pizza, Italian', 4)]

### 5.6: Top 10 High-Rated Restaurant Categories
#### Shows high-rated restaurant types, helping you decide which categories might attract quality-conscious customers.

In [32]:
%%sql
SELECT TOP 10 categories, AVG(stars) AS avg_stars
FROM business
WHERE city = 'Tampa' AND categories LIKE '%Restaurant%'
GROUP BY categories
ORDER BY avg_stars DESC;

 * mssql+pyodbc://anishp5admin:***@anishp5-server.database.windows.net/MiniProject-1?connect_timeout=30&driver=ODBC+Driver+17+for+SQL+Server
Done.


[('American (Traditional), Italian, Personal Chefs, Restaurants, Event Planning & Services, Sandwiches, Caterers', 5.0),
 ('Bakeries, Food, Vegan, Cupcakes, Cafes, Donuts, Restaurants, Desserts', 5.0),
 ('Breakfast & Brunch, Restaurants, Dominican, Caribbean, Food, Coffee & Tea', 5.0),
 ('Cafes, Cuban, Restaurants, Nightlife, Bars', 5.0),
 ('Food Stands, Restaurants, Food Trucks, Specialty Food, Food', 5.0),
 ('Food Trucks, Restaurants, Italian, Caterers, Event Planning & Services, Food', 5.0),
 ('Food, Restaurants, Specialty Food, Russian', 5.0),
 ('Food, Street Vendors, Japanese, Restaurants', 5.0),
 ('Gluten-Free, Poke, Food, Restaurants, Hawaiian', 5.0),
 ('Hot Dogs, Vegetarian, Burgers, Restaurants, Food Stands, Sandwiches', 5.0)]

### 5.7: Commonly Reviewed Restaurants in the City
#### Identifies popular restaurants based on review count, which can signal high customer interest or strong marketing

In [33]:
%%sql
SELECT TOP 10 name, review_count
FROM business
WHERE city = 'Tampa' AND categories LIKE '%Restaurant%'
ORDER BY review_count DESC;

 * mssql+pyodbc://anishp5admin:***@anishp5-server.database.windows.net/MiniProject-1?connect_timeout=30&driver=ODBC+Driver+17+for+SQL+Server
Done.


[('Jacksons Bistro Bar & Sushi', 1115),
 ('Besito Mexican Restaurant', 1092),
 ('American Social', 768),
 ('Rusty Pelican - Tampa', 745),
 ('Seminole Hard Rock Hotel & Casino Tampa', 739),
 ('Topgolf', 734),
 ('Yah Mon', 697),
 ('Brick House Tavern + Tap', 561),
 ('The Bricks', 548),
 ('Ava', 544)]

### 5.8: Number of Open vs. Closed Restaurants
#### Shows the ratio of open to closed restaurants, potentially indicating market saturation or challenges.

In [36]:
%%sql
SELECT 
    is_open,
    COUNT(*) AS num_restaurants,
    ROUND((COUNT(*) * 100.0 / (SELECT COUNT(*) FROM business WHERE city = 'Tampa' AND categories LIKE '%Restaurant%')),2) AS percentage
FROM business
WHERE city = 'Tampa' AND categories LIKE '%Restaurant%'
GROUP BY is_open;



 * mssql+pyodbc://anishp5admin:***@anishp5-server.database.windows.net/MiniProject-1?connect_timeout=30&driver=ODBC+Driver+17+for+SQL+Server
Done.


[(0, 174, Decimal('34.940000000000')), (1, 324, Decimal('65.060000000000'))]

### 5.9: Identify Areas in Tampa with High-Rated Restaurants
#### Pinpoints postal codes with high-rated restaurants, helping in selecting an area known for quality dining options.

In [38]:
%%sql
SELECT TOP 10 postal_code, ROUND(AVG(stars),2) AS avg_stars
FROM business
WHERE city = 'Tampa' AND categories LIKE '%Restaurant%'
GROUP BY postal_code
ORDER BY avg_stars DESC;

 * mssql+pyodbc://anishp5admin:***@anishp5-server.database.windows.net/MiniProject-1?connect_timeout=30&driver=ODBC+Driver+17+for+SQL+Server
Done.


[('33616', 4.5),
 ('33637', 4.17),
 ('33621', 4.0),
 ('33605', 3.95),
 ('33606', 3.9),
 ('33603', 3.88),
 ('33604', 3.77),
 ('33615', 3.77),
 ('33619', 3.73),
 ('33624', 3.72)]

### 5.10: Average Rating by Restaurant Attributes
#### Shows which attributes (e.g., "Good for Groups," "Outdoor Seating") are associated with higher ratings, providing insight into customer preferences.

In [41]:
%%sql
SELECT 
    ROUND(AVG(CASE WHEN attributes LIKE '%"BusinessAcceptsCreditCards": "True"%' THEN stars END),2) AS avg_stars_credit_cards,
    ROUND(AVG(CASE WHEN attributes LIKE '%"RestaurantsDelivery": "True"%' THEN stars END),2) AS avg_stars_delivery,
    ROUND(AVG(CASE WHEN attributes LIKE '%"OutdoorSeating": "True"%' THEN stars END),2) AS avg_stars_outdoor_seating,
    ROUND(AVG(CASE WHEN attributes LIKE '%"RestaurantsTakeOut": "True"%' THEN stars END),2) AS avg_stars_takeout,
    ROUND(AVG(CASE WHEN attributes LIKE '%"BikeParking": "True"%' THEN stars END),2) AS avg_stars_bike_parking
FROM business
WHERE city = 'Tampa' AND categories LIKE '%Restaurant%';


 * mssql+pyodbc://anishp5admin:***@anishp5-server.database.windows.net/MiniProject-1?connect_timeout=30&driver=ODBC+Driver+17+for+SQL+Server
Done.


[(3.66, 3.56, 3.82, 3.68, 3.76)]

## PART 6: Conclusion for Setting Up a Restaurant in Tampa

#### Why Tampa?
Tampa ranks second in terms of the number of top-rated restaurants, only behind Philadelphia, and has an impressive average restaurant rating of 3.67, the highest among the cities we analyzed. This indicates that Tampa has a thriving restaurant scene, with a strong culture of dining out and a competitive but appealing environment. The average number of fans per user in Tampa is 39, which is quite significant and suggests the presence of an engaged local foodie community that can drive word-of-mouth promotion.

#### Popular Cuisine Preferences
Our data shows that certain categories, such as Chinese and Pizza, are particularly popular, with multiple businesses in these categories gaining significant traction. For a new restaurant to differentiate itself, we can either tap into these popular cuisines or consider a unique fusion approach, like offering Chinese-Italian fusion dishes, to attract curious and adventurous diners.

#### Strategic Neighborhood Selection
Neighborhood analysis highlights postal codes 33616 and 33637 as prime areas, with average restaurant ratings of 4.5 and 4.17, respectively. These areas are hubs for highly rated establishments, making them ideal for launching a new restaurant that targets quality-seeking customers. Postal codes like 33605 and 33606, with ratings just shy of 4.0, also represent opportunities where a well-executed restaurant concept could quickly rise to prominence.

#### Key Restaurant Features for Success
The analysis of restaurant attributes reveals that restaurants with outdoor seating have the highest average rating of 3.82. Takeout and bike parking options also score highly, with ratings of 3.68 and 3.76, respectively. This suggests that incorporating these features—such as a well-designed outdoor dining area and efficient takeout services—could enhance customer satisfaction and attract a broader audience. Additionally, given Tampa’s warm climate and active community, outdoor seating and bike parking would cater well to local preferences.

#### Operational Insights
The ratio of open to closed restaurants in Tampa is favorable, with 65.06% of establishments currently operating. This indicates a relatively stable market environment where a well-conceived restaurant has a good chance of succeeding, provided it delivers quality and caters to local preferences. It also highlights the importance of ongoing adaptation to maintain relevance and operational sustainability.
